warehouse/manage/views.py

```diff
@@ -616,6 +616,70 @@ def delete_webauthn(self):
         return HTTPSeeOther(self.request.route_path("manage.account"))
 
 
+@view_defaults(
+    uses_session=True,
+    require_csrf=True,
+    require_methods=False,
+    permission="manage:user",
+    http_cache=0,
+    has_translations=True,
+)
+class ProvisionRecoveryCodesViews:
+    def __init__(self, request):
+        self.request = request
+        self.user_service = request.find_service(IUserService, context=None)
+
+    @view_config(
+        request_method="GET",
+        route_name="manage.account.recovery-codes.generate",
+        renderer="manage/account/recovery_codes-provision.html",
+    )
+    def recovery_codes_generate(self):
+        if self.user_service.has_recovery_codes(self.request.user.id):
+            self.request.session.flash(
+                _("Recovery codes already generated"), queue="error"
+            )
+            return HTTPSeeOther(self.request.route_path("manage.account"))
```

Comment:
IMO it would be more user-friendly to still render this page, and just render this warning instead of the codes, rather than flashing/redirecting. Could also guide the user to re-generating as that might be what they want to do.